File: src/main/java/com/donrobo/fpbg/blueprint/Direction.java
package com.donrobo.fpbg.blueprint;

import com.donrobo.fpbg.data.Int2;

public enum Direction {

    UP(4),
    DOWN(0),
    LEFT(2),
    RIGHT(6);

    private final int directionValue;

    Direction(int directionValue) {
        this.directionValue = directionValue;
    }

    public int getDirectionValue() {
        return directionValue;
    }

    public Direction reverseDirection() {
        switch (this) {
            case UP:
                return DOWN;
            case DOWN:
                return UP;
            case LEFT:
                return RIGHT;
            case RIGHT:
                return LEFT;
            default:
                throw new RuntimeException("New direction added?");
        }
    }

    public Int2 move(Int2 start, int distance) {
        int actualX = start.getX();
        int actualY = start.getY();

        switch (this) {
            case UP:
                actualY -= distance;
                break;
            case DOWN:
                actualY += distance;
                break;
            case LEFT:
                actualX -= distance;
                break;
            case RIGHT:
                actualX += distance;
                break;
        }

        return new Int2(actualX, actualY);
    }
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/Direction.kt
package com.donrobo.fpbg.blueprint

import com.donrobo.fpbg.data.Int2

enum class Direction(val directionValue: Int) {

    UP(4),
    DOWN(0),
    LEFT(2),
    RIGHT(6);

    val reversed: Direction get() = when (this) {
        UP -> DOWN
        DOWN -> UP
        LEFT -> RIGHT
        RIGHT -> LEFT
        else -> throw RuntimeException("New direction added?")
    }

    fun move(start: Int2, distance: Int = 1): Int2 {
        var actualX = start.x
        var actualY = start.y

        when (this) {
            UP -> actualY -= distance
            DOWN -> actualY += distance
            LEFT -> actualX -= distance
            RIGHT -> actualX += distance
        }

        return Int2(actualX, actualY)
    }

    fun rotateCW(count: Int): Direction {
        return if (count == 0)
            this
        else if (count <= 3)
            when (this) {
                UP -> RIGHT
                RIGHT -> DOWN
                DOWN -> LEFT
                LEFT -> UP
            }.rotateCW(count - 1)
        else
            throw IllegalArgumentException("Only count 1-3 are allowed!")

    }


    companion object {
        fun calculateDirection(from: Int2, to: Int2): Direction {
            assert((from.x == to.x) != (from.y == to.y))
            if (from.x < to.x) {
                return Direction.RIGHT
            } else if (from.x > to.x) {
                return Direction.LEFT
            } else if (from.y < to.y) {
                return Direction.DOWN
            } else if (from.y > to.y) {
                return Direction.UP
            } else {
                throw RuntimeException("Shouldn't happen")
            }
        }
    }
}

fun Int2.move(direction: Direction, distance: Int = 1): Int2 {
    return direction.move(this, distance)
}